### ISSUE TITLE
test relies to external repository

### ISSUE THREAD

**Issue by vaab** [audreyr/cookiecutter#434]
I'm not sure having extensive test actually needing to use web connection is a good idea. I'm currently in China, and connections towards western world are not so easy. Even with a VPN, I can't launch the test suite without having several test failing while cloning (git clone or hg clone).

I also noticed that travis as difficulties itself to clone all these (https://travis-ci.org/0k/cookiecutter/builds/56044642). (along with strange travis issues not finding ``py33`` or ``flake8``) 

And any bad push to one of these external repository will have the consequence to break tests from cookiecutter.
At last, having these external dependency raises the whole test time, which is never a good thing as it is an incitation to run tests less often.

Basically, there probably shouldn' t be even 1 test having to rely to external sites, you could test git and mercurial locally, and rely on fixed version of some reference templates.

My 2 cents.

**Comment by pydanny** [comment]
The problem with mocks is that they are... well... mocks. If we replace our current technique of integration tests with them, how do we know if changes borked anything? Cookiecutter does a ton of network access stuff, as you so aptly pointed out.

Now that said, what we could do is implement `skipif` controlled by an environment variable on integration tests. That way, if you don't want to test that side of things, you can skip it.

**Comment by vaab** [comment]
Why are you mentioning mocks ? Having a frozen version (copied in a selected state in the test suite) of selected templates is not what I would have called a mock (but well, tests and vocabulary around them is kind of variable). 
I don't understand what using this would change related to your question (namely the "If we replace our current technique of integration tests with them, how do we know if changes borked anything?"). To put it clearly: you will cover the exact same functional test minus what you are not supposed to test: You are certainly NOT supposed to:
- test connection of the testing computer towards internet, 
- availability of various providers, 
- other project's (the templates) last version's conformity to current cookiecutter.
- current VCS tools (git, hr) ability to clone from distant URL
You are testing way too much than expected and that it has a direct consequence on your testing suite.

**Comment by hackebrot** [comment]
Well we already use `skipif` markers for certain tests. :relaxed:

You can set an according environment variable **DISABLE_NETWORK_TESTS**, please see  https://github.com/audreyr/cookiecutter/blob/master/tests/skipif_markers.py#L23

The markers are used in the following test modules:
```
tests/test_vcs_prompt.py
tests/test_cookiecutter_repo_arg.py
tests/test_vcs.py
tests/test_more_cookiecutters.py
tests/test_cookiecutters.py
```
We could also set up our own marker that deselects tests if desired and have an own target for this in the make file (https://pytest.org/latest/example/markers.html#mark-examples).

**Comment by vaab** [comment]
IMHO, then these tests should be skipped always, as they regularly make the current integration fails for network reasons, availability of providers, and will regularly make the test fails depending on others packages conformity.

And removing test is not necessary, most of them could be only adapted for non-networking. I'm not speaking of mock, but more a simple local copy of selected version of selected templates. The functional coverage on cookiecutter won't change. 

Package conformity should be tested on the package side (as I did with https://github.com/0k/cookiecutter-odoo ). Breaking the templates should not break cookiecutters test suite.

Doesn't all this sound like evidence for you ? I'm quite puzzled that it doesn't. I might have missed something important there, please take time to explain me what I've might have overlooked.

**Comment by michaeljoseph** [comment]
@vaab I'm :+1: on this idea.
Something like shipping known good versions of cookiecutter [cookiecutters](https://github.com/audreyr/cookiecutter/issues/430) as test data and setting them up for local cloning?
Would you like to take a stab at this?

**Comment by vaab** [comment]
Yes of course. I just wanted to collect the heat on the subject... I'll try to do something in the following week if I have some spare time/
